baseapp/baseapp.go

```diff
@@ -623,16 +623,40 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte, tx sdk.Tx) (result sdk
 		return err.Result()
 	}
 
-	// run the ante handler
+	// Keep the state in a transient CacheWrap in case executing the ante handler
+	// or processing the messages fails.
+	msCache := getState(app, mode).CacheMultiStore()
+	if msCache.TracingEnabled() {
+		msCache = msCache.WithTracingContext(
+			sdk.TraceContext(
+				map[string]interface{}{
+					"txHash": cmn.HexBytes(tmhash.Sum(txBytes)).String(),
+				},
+			),
+		).(sdk.CacheMultiStore)
+	}
+
+	// create a new context with the transient CacheWrap multi-store
+	ctx = ctx.WithMultiStore(msCache)
+
+	// Execute the ante handler. If the ante handler fails and thus needs to abort,
+	// state modifications will not be persisted due to the transient CacheWrap
+	// multi-store.
 	if app.anteHandler != nil {
 		newCtx, result, abort := app.anteHandler(ctx, tx, (mode == runTxModeSimulate))
 		if abort {
 			return result
 		}
+
 		if !newCtx.IsZero() {
 			ctx = newCtx
 		}
 
+		// if during deliverTx the ante handler execution was successful, persist state
+		if mode == runTxModeDeliver {
```

Comment:
Should we also persist changes to the checktx state? Presently we do, so this will change that behavior.

For example, this would prevent an account with only enough fees for one transaction from getting two past `CheckTx` successfully.

Reply:
I'm a bit confused here (partially why I also think we need much better documentation), so maybe you can help me out here?

If we persist ante handler state transitions for both checkTx **and** deliverTx, then state changes like nonce incrementation and fee deduction happen twice? This isn't immediately obvious.

For example, `TestGaiaCLISendGenerateSignAndBroadcast` was failing before I had the `runTxModeDeliver` check because it expected the nonce to be 1 when it was really 2.

Reply:
No, they (should) use separate states - `app.deliverState`, `app.checkState`.

Reply:
Ok, after further digging through BaseApp, I've noticed that checkTx state is not actually persisted between txs, but is persisted between potential checkTx calls. So the current logic will not do.

Instead, we will only use this cached context during deliverTx and checkTx logic will remain unchanged. I'll make note of this in the documentation.